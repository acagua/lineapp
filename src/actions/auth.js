import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebaseConfig'
import { finishLoading, startLoading } from "./ui";
import Swal from 'sweetalert2';
import { db } from '../firebase/firebaseConfig';
import { loadUser } from "../helpers/loadUser";

//return si es asincrono, sin return si no lo es
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        
        dispatch( startLoading() );

        firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                
                dispatch( login(user.uid) );
                dispatch( finishLoading());
            }).catch( ({ message }) => {
                dispatch( finishLoading()); // podría ser en el finally pero complica el manejo de error con swal
                Swal.fire('Error', message, 'error' );
            })
    }
}

export const startRegister = (email, password, name, phone) => {
    return (dispatch) => {
        dispatch( startLoading() );
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({ user }) => {

                await user.updateProfile({
                    displayName: name
                });
                db.collection("users").doc(user.uid)
                .set({
                    name,
                    phone,
                })

                dispatch (login (user.uid));
                dispatch( finishLoading());

            })
            .catch ( e => {
                dispatch( startLoading() );
                Swal.fire('Error', e.message, 'error' );
            });
    }
}

export const startLoginGoogle = (email, password) => { // si es asínctrona se hace el return (dispatch)
    return (dispatch) => {
        firebase.auth().signInWithPopup( googleAuthProvider )
            .then( ({ user }) => {
                dispatch (login (user.uid));
            } );
    }
}

export const login = (uid) => ({
    type: types.login,
    payload: {
        uid
    }
});

export const startLoadingUser = ( uid ) =>{
    return async ( dispatch ) => {
        const user = await loadUser ( uid );
        dispatch( setLoadedUser ( user ) );
    }
}
export const setLoadedUser = ( user ) => ({
    type: types.authLoadUser,
    payload: user
})


export const startLogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut();

        dispatch ( logout() );

        // dispatch ( noteLogout() );
    }
}
export const logout = () => ({
    type: types.logout,
});