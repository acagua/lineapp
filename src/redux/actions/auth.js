import { types } from "../types/types";
import { db, firebase } from '../../firebase/firebaseConfig'
import { finishLoading, startLoading } from "./ui";
import { loadUser } from "../../helpers/loadUser";
import Swal from 'sweetalert2';
import { companiesLogout } from "./companies";
import { linesLogout } from "./lines";

//return si es asincrono, sin return si no lo es
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        
        dispatch( startLoading() );

        firebase.auth().signInWithEmailAndPassword(email, password)
            .then(async({ user }) => {
                
                await dispatch(startLoadingUser(user.uid));
                // dispatch( login(user.uid) );
                dispatch( finishLoading());
            }).catch( ({ message }) => {
                dispatch( finishLoading()); // podría ser en el finally pero complica el manejo de error con swal
                Swal.fire('Error', message, 'error' );
            })
    }
}

export const startRegister = (email, password, name, phone, role) => {
    return (dispatch) => {
        dispatch( startLoading() );
        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({ user }) => {

                await user.updateProfile({
                    displayName: name
                });
                db.collection("users").doc(user.uid)
                .set({
                    name,
                    phone,
                    role, 
                })

                await dispatch (startLoadingUser(user.uid));
                // dispatch (login (user.uid));
                dispatch( finishLoading());

            })
            .catch ( e => {
                dispatch( startLoading() );
                Swal.fire('Error', e.message, 'error' );
            });
    }
}

// export const startLoginGoogle = (email, password) => { // si es asínctrona se hace el return (dispatch)
//     return (dispatch) => {
//         firebase.auth().signInWithPopup( googleAuthProvider )
//             .then( ({ user }) => {
//                 dispatch (login (user.uid));
//             } );
//     }
// }

// export const login = (uid) => ({
//     type: types.login,
//     payload: {
//         uid
//     }
// });

export const startLoadingUser = ( uid ) =>{
    return async ( dispatch ) => {
        const user = await loadUser ( uid );
        dispatch( setLoadedUser (uid, user) );
    }
}
export const setLoadedUser = ( uid, user ) => ({
    type: types.authLoadUser,
    payload: {uid, user}
})


export const startLogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut();

        dispatch ( logout() );
        dispatch( companiesLogout());
        dispatch( linesLogout());

    }
}
export const logout = () => ({
    type: types.logout,
});