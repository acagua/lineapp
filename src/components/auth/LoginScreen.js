import React, { useState } from 'react'
import { useDispatch} from 'react-redux';

import validator from "validator";
// import Swal from "sweetalert2";
import { Link } from 'react-router-dom';
import { startLoginEmailPassword } from '../../redux/actions/auth';

const errorInitState = {
  error: false,
  errMessage: "",
}

export const LoginScreen = () => {
    const [formValues, setFormValues] = useState({
        email: "",
        password: ""
    });

    const dispatch = useDispatch();
    
    const [loginError, setLoginError] = useState(errorInitState);

    const { email, password } = formValues;
    const { error, errMessage } = loginError;

    const handleInputChange = ( { target } ) =>{
		setFormValues({
			...formValues,
			[target.name]: target.value
		})
	}

    const handleLogin = (e) => {
        e.preventDefault();
        if (formIsValid()) {
            //Swal.fire("Login", "Successful login!", "success");
            dispatch( startLoginEmailPassword(email, password ) );
        }
    };

    const formIsValid = () => {
        if (!validator.isEmail(email)) {
            setLoginError(
            {
                error:true,
                errMessage:'Email is not valid'
            });
            return false;
        } else if (password.length < 5) {
            setLoginError(
            {
                error:true,
                errMessage:'Password must be at least 6 characters long'
            });
            return false;
        }
        setLoginError(errorInitState);

        return true;
    };
    return (
        <div className="auth__main">
            <div className="auth__form-container">
                <h1 className="auth__title mb-5"> Login </h1>
                <form className="mb-3" onSubmit={ handleLogin }>
                    {
                        (error) && 
                        (
                            <div id="error" className="auth__alert-error mb-2">
                                { errMessage }
                            </div>
                        )
                    }
                    <input 
                        type="text"
                        id="email"
                        placeholder="Email"
                        name="email"
                        className={`auth__input form-control`}
                        autoComplete="off"
                        value = { email }
                        onChange = { handleInputChange }
                    />
                    <input 
                        type="password"
                        id="password"
                        placeholder="Password"
                        name="password"
                        // className={`auth__input form-control ${ !validPassword && 'is-invalid' }`}
                        className={`auth__input form-control`}
                        value = { password }
                        onChange = { handleInputChange }
                    />
                    <button
                        type="submit"
                        className="btn btn-primary btn-block mt-3"
                        disabled = { !( email.length > 0 &&  password.length > 0) }
                    >
                        Login
                    </button>
                    
                </form>
                <Link 
                    to="/auth/register"
                    className="link"
                >
                    Create new account
                </Link>
            </div>
        </div>
    )
}
