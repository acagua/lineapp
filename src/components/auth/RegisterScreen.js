import React, { useState } from 'react'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';

import { useForm } from '../../hooks/useForm'
import { startRegister } from '../../actions/auth';

const errorInitState = {
    error: false,
    errMessage: "",
  }
  const initialForm = {
      name: 'Andres',
      email: 'andrescagua@gmail.com',
      phone: '3138891481',
      password: '123456',
      password2: '123456',
    }
    
    export const RegisterScreen = () => {
        
        
    const dispatch = useDispatch();
    
    const { loading } = useSelector( state => state.ui );
    
    const [formError, setFormError] = useState(errorInitState);

    const { error, errMessage } = formError;

    const [ formValues, handleInputChange ] = useForm(initialForm);

    const {name,email, phone, password, password2} = formValues;
   
    const handleRegister = (e) => {
        e.preventDefault();
        if(isFormValid()){
            
            dispatch(startRegister(email, password, name, phone));
        }
    }

    const isFormValid = () => {
        
        if(name.trim().length === 0){
            setFormError(
            {
                error:true,
                errMessage:'Please enter your name'
            });
            return false;
        } else if ( !(/^\d+$/.test(phone)) ){
            setFormError(
            {
                error:true,
                errMessage:'Phone must only have digits'
            });
            return false;
        } else if (phone.length < 10 ){
            setFormError(
            {
                error:true,
                errMessage:'Phone must have 10 digits'
            });
            return false;
        }else if (!validator.isEmail(email)){
            setFormError(
            {
                error:true,
                errMessage:'Email is not valid'
            });
            return false;
        } else if ( password.length < 5){
            setFormError(
            {
                error:true,
                errMessage:'Password must be at least 6 characters long'
            });
            return false;
        } else if ( password !== password2){
            setFormError(
            {
                error:true,
                errMessage:'Passwords are not the same'
            });
            return false;
        }
        
        setFormError(errorInitState);
        
        return true;
    }

    return (
        <div className="auth__main">
            <div className="auth__form-container">
                <h1 className="auth__title mb-5"> New account </h1>
                <form className="mb-3" onSubmit={ handleRegister }>
                    {
                        (error) && 
                        (
                            <div id="error" className="auth__alert-error mb-2">
                                { errMessage }
                            </div>
                        )
                    }
                    <input 
                        type="text"
                        placeholder="Name"
                        name="name"
                        className={`auth__input form-control`}
                        autoComplete="off"
                        value = { name }
                        onChange = { handleInputChange }
                        required
                    />
                    {/* <input 
                        type="number"
                        placeholder="Phone"
                        name="phone"
                        className={`auth__input form-control`}
                        autoComplete="off"
                        // value = { email }
                        // onChange = { handleInputChange }
                        required
                    /> */}
                    <input 
                        type="tel"
                        placeholder="Phone"
                        name="phone"
                        className={`auth__input form-control`}
                        autoComplete="off"
                        value = { phone }
                        onChange = { handleInputChange }
                        required
                    />
                    <input 
                        type="text"
                        placeholder="Email"
                        name="email"
                        className={`auth__input form-control`}
                        autoComplete="off"
                        value = { email }
                        onChange = { handleInputChange }
                        required
                    />
                    <input 
                        type="password"
                        placeholder="Password"
                        name="password"
                        autoComplete="off"
                        // className={`auth__input form-control ${ !validPassword && 'is-invalid' }`}
                        className={`auth__input form-control`}
                        value = { password }
                        onChange = { handleInputChange }
                        required
                    />
                    <input 
                        type="password"
                        placeholder="Confirm Password"
                        name="password2"
                        autoComplete="off"
                        // className={`auth__input form-control ${ !validPassword && 'is-invalid' }`}
                        className={`auth__input form-control`}
                        value = { password2 }
                        onChange = { handleInputChange }
                        required
                    />
                    <button
                        type="submit"
                        className="btn btn-primary btn-block mt-3"
                        disabled = { loading || !( email.length > 0 &&  password.length > 0) } 
                    >
                        Sign Up
                    </button>
                    
                </form>
                <Link 
                    to="/auth/login"
                    className="link"
                >
                    Already have an account?
                </Link>
            </div>
        </div>
    )
}
