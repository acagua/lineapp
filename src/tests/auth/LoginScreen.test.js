import '@testing-library/jest-dom';
import { LoginScreen } from '../../components/auth/LoginScreen'
import { mount, shallow } from 'enzyme'
//jest.mock('../../components/auth/LoginScreen');

describe('Pruebas sobre componente auth/LoginScreen', () => {

    const wrapper = mount ( <LoginScreen />);
    
    test('Se muestra el componente correctamente', () => {
        expect(wrapper).toMatchSnapshot();
        
    });

    test ('El formulario se encuentra vacío por defecto', () => {
        
        const loginButtonDisabled = wrapper.find('button').prop('disabled');
        const email = wrapper.find('#email').prop('value');
        const password = wrapper.find('#password').prop('value');

        expect(email).toBe('');
        expect(password).toBe('');
        expect(loginButtonDisabled).toBe(true);
    }); 

    test('Permite el ingreso del correo correctamente', () => {
        const email = wrapper.find('#email');
        expect(email.prop('value')).toBe('');

        email.simulate('change', { target: { name:'email', value: 'random-text' } });
        expect(wrapper.find('#email').prop('value')).toBe('random-text');
        
    });

    test('Permite el ingreso de la contraseña correctamente', () => {
        
        const password = wrapper.find('#password');
        expect(password.prop('value')).toBe('');
        
        password.simulate('change', { target: { name:'password', value: 'random-text' } });
        expect(wrapper.find('#password').prop('value')).toBe('random-text');
        
    });


    test('Habilita el botón de Login si los campos tienen datos', () => {
        
        const loginButton = wrapper.find('button').prop('disabled');
        expect(loginButton).toBe(false);

    });



    test('Muestra mensaje de error cuando el correo no es válido', () => {
        
        const email = wrapper.find('#email');
        email.simulate('change', { target: { name:'email', value: 'test@something' } });
        const loginButton = wrapper.find('button');
        loginButton.simulate("click");

        // expect(loginButton).toBe('Email is not valid');
        
        //const error =  wrapper.find('#error').props();

        //expect(error).toBe('Email is not valid');
        expect(wrapper.contains('Email is not valid')).toBe(true);

    });
    test('Debería mostrar mensaje de error sobre contraseña incorrecta', () => {
        
        const a = 1;
        const b = 1;

        expect(a).toBe(b);

    });
});
