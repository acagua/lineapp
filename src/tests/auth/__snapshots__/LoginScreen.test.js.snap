// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre componente auth/LoginScreen Se muestra el componente correctamente 1`] = `
<div
  className="auth__main"
>
  <div
    className="auth__form-container"
  >
    <h1
      className="auth__title mb-5"
    >
       Login 
    </h1>
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="auth__input form-control"
        id="email"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
      <input
        className="auth__input form-control"
        id="password"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
      <button
        className="btn btn-primary btn-block mt-3"
        disabled={true}
        type="submit"
      >
        Login
      </button>
      <div
        className="auth__alert-error mt-3"
        id="error"
      >
        Email is not valid
      </div>
    </form>
  </div>
</div>
`;
